flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointMetadataOutputStream.java

```diff
@@ -109,13 +119,19 @@ public void close() {
 
             try {
                 out.close();
-                fileSystem.delete(metadataFilePath, false);
+                if (!isRecoverableStream(out)) {
+                    fileSystem.delete(metadataFilePath, false);
+                }
             } catch (Throwable t) {
                 LOG.warn("Could not close the state stream for {}.", metadataFilePath, t);
             }
         }
     }
 
+    private boolean isRecoverableStream(FSDataOutputStream out) {
```

Comment:
It seems relaying on conditions and branches in different places might complicate the implementation. Might you introduce an interface for the two types of implementation? like

```
interface MetadataOutputStreamBackend  {
     
    FSDataOutputStream getOutput();

    void commit();

    void close();
}
```

and have two implementations ?

Reply:
Sounds great!